lib/resources/postgres.rb

```diff
@@ -18,29 +19,24 @@ def initialize
         # Debian allows multiple versions of postgresql to be
         # installed as well as multiple "clusters" to be configured.
         #
-        version = version_from_dir('/etc/postgresql')
-        cluster = cluster_from_dir("/etc/postgresql/#{version}")
-        @conf_dir = "/etc/postgresql/#{version}/#{cluster}"
-        @data_dir = "/var/lib/postgresql/#{version}/#{cluster}"
+        @version = version_from_psql || version_from_dir('/etc/postgresql')
+        @cluster = cluster_from_dir("/etc/postgresql/#{@version}")
+        @conf_dir = "/etc/postgresql/#{@version}/#{@cluster}"
+        @data_dir = "/var/lib/postgresql/#{@version}/#{@cluster}"
       elsif os.redhat?
-        #
-        # /var/lib/pgsql/data is the default data directory on RHEL6
-        # and RHEL7. However, PR #824 explicitly added version-based
-        # directories. Thus, we call #version_from_dir unless it looks
-        # like we are using unversioned directories.
-        #
-        # TODO(ssd): This has the potential to be noisy because of the
-        # warning in version_from_dir. We should determine which case
-        # is more common and only warn in the less common case.
-        #
-        version = if inspec.directory('/var/lib/pgsql/data').exist?
-                    warn 'Found /var/lib/pgsql/data. Assuming postgresql install uses un-versioned directories.'
-                    nil
-                  else
-                    version_from_dir('/var/lib/pgsql/')
-                  end
+        @version = version_from_psql
+        if @version.nil?
+          if inspec.directory('/var/lib/pgsql/data').exist?
+            warn 'PSQL did not return a version number and found /var/lib/pgsql/data.
```

Comment:
I'd personally get rid of this warning statement. I'm not sure it adds any value to the user.

If we choose to leave it, we should change the wording because an end-user shouldn't need to worry about what a "nil" is, etc.

Suggested wording:

```
Unable to determine PostgreSQL version: psql did not return a version number and unversioned data directories were found.
```

But my recommendation is to eliminate it. See below for another commend on exposing the detection mechanism to the users if that's the problem we're trying to solve.